rkt/containers.go

```diff
@@ -0,0 +1,165 @@
+)
+
+type containerHandle struct {
+	uuid       string
+	lock       *lock.DirLock
+	isExited   bool
+	isGarbage  bool
+	isDeleting bool
+}
+
+type includeMask byte
+
+const (
+	includeContainersDir includeMask = 1 << iota
+	includeGarbageDir
+)
+
+// getContainerHandles returns a slice of ContainerHandle structs representing the current state of all containers
+func getContainerHandles(include includeMask) ([]*containerHandle, error) {
+	ls, err := getContainers(include)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get containers: %v", err)
+	}
+
+	var cs []*containerHandle
+	for _, uuid := range ls {
+		state, err := getContainerHandle(uuid)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error getting state of container %s, ignoring\n", uuid)
+			continue
+		}
+		cs = append(cs, state)
+	}
+
+	return cs, nil
+}
+
+// getContainerHandle returns a ContainerHandle struct representing the given container
+func getContainerHandle(uuid string) (*containerHandle, error) {
+	s := &containerHandle{}
+
+	l, err := lock.NewLock(filepath.Join(containersDir(), uuid))
+	if err == lock.ErrNotExist {
+		l, err = lock.NewLock(filepath.Join(garbageDir(), uuid))
+		s.isGarbage = true
+		s.isExited = true
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error opening container %q: %v", uuid, err)
+	}
+
+	if err = l.TrySharedLock(); err != nil {
+		if err != lock.ErrLocked {
+			l.Close()
+			return nil, fmt.Errorf("unexpected lock error: %v", err)
+		}
+		if s.isGarbage {
+			s.isDeleting = true
+		}
+	} else {
+		l.Unlock()
+		s.isExited = true
+	}
+
+	s.uuid = uuid
+	s.lock = l
+
+	return s, nil
+}
+
+// freeContainerHandle closes the associated lock of a container state opened via getContainerHandle
+func freeContainerHandle(s *containerHandle) {
+	s.lock.Close()
+}
+
+// path returns the path to the container where it was opened
+func (ch *containerHandle) path() string {
+	if ch.isGarbage {
+		return ch.garbagePath()
+	} else {
+		return ch.containersPath()
+	}
+}
+
+// garbagePath returns the path to the container where it would be in the garbageDir
+func (ch *containerHandle) garbagePath() string {
+	return filepath.Join(garbageDir(), ch.uuid)
+}
+
+// containersPath returns the path to the container where it would be in the containersDir
+func (ch *containerHandle) containersPath() string {
+	return filepath.Join(containersDir(), ch.uuid)
+}
+
+// getContainers returns a list of container uuids in string form
+func getContainers(include includeMask) ([]string, error) {
+	var (
+		cs  []string
+		err error
+	)
+
+	if include&includeContainersDir != 0 {
+		cs, err = getContainersFromDir(containersDir())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if include&includeGarbageDir != 0 {
+		gcs, err := getContainersFromDir(garbageDir())
+		if err != nil {
+			return nil, err
+		}
+		cs = append(cs, gcs...)
+	}
+
+	return cs, nil
+}
+
+// getContainersFromDir returns a list of container uuids in string form from a specific directory
```

Comment:
uuids are always strings right? Maybe we simply the comment?

```
// getContainersFromDir returns a list of container uuids in string from the given directory.
```

Notice the addition of the `.`